feat: travel schedules to schedule timezone changes

CarinaWolli opened the pull request:
## What does this PR do?

New feature to add travel schedules for scheduling a timezone change. 

Loom showing availability with travel schedules: https://www.loom.com/share/29d8500385fb4ff78e025021ad5710ea

Fixes #6790

## Type of change

- New feature (non-breaking change which adds functionality)

## How should this be tested?

To run locally change your system timezone to utc time.

- Create travel schedules 
- test availability 
- test CRON API endpoint `/api/cron/changeTimeZone`
    - change system timezone to right after travel schedule started (midnight of user timezone)
    - check if timezone changed & check if timezone of default schedule changed

Commit:
add Travel Schedule modal

Commit:
UI to list schedules

Commit:
Merge branch 'main' into feat/schedule-timezone-change

Commit:
set shouldDirty

Commit:
add backend  to add new travelSchedule

Commit:
implement deleting a schedule

Commit:
Merge branch 'main' into feat/schedule-timezone-change

Commit:
check if schedule is overlapping

Commit:
WIP

Commit:
Merge branch 'main' into feat/schedule-timezone-change

Commit:
fix finding overlapping travel schedule

Commit:
only use travelSchedule when default availability is used

Commit:
Merge branch 'main' into feat/schedule-timezone-change

Commit:
adjust date overrides to timezone schedule

Commit:
first version of changeTimeZone cron job

Commit:
Merge branch 'main' into feat/schedule-timezone-change

Commit:
fix tests by adding travelSchedules

Commit:
fixes for cron job api call

Commit:
add first unit test for travel schedules

Commit:
improve unit tests

Commit:
Merge branch 'main' into feat/schedule-timezone-change

Commit:
add migration

Commit:
fix type error

Commit:
fix collective-scheduling test

Commit:
clean up cron job

Commit:
code clean up

Commit:
code clean up

Commit:
show timezone from travel schedule for date override

PeerRich commented:
ouhhhh exciting!!

Commit:
Merge branch 'main' into feat/schedule-timezone-change

Commit:
add date override tests

Commit:
show tz on date override only in default schedule

Commit:
fix deleting old schedules

Commit:
minor fixes in cron api handler

CarinaWolli commented on packages/features/schedules/components/DateOverrideList.tsx:
```diff
@@ -72,7 +78,17 @@ const DateOverrideList = ({
               item.ranges.map((range, i) => (
                 <p key={i} className="text-subtle text-xs">
-                  {timeSpan(range)}
+                  {`${timeSpan(range)} ${
```
Show timezone to date override if it is within a travel schedule (only on default schedule): 
![Screenshot 2024-03-04 at 6 18 51 PM](https://github.com/calcom/cal.com/assets/30310907/11adb650-f919-47b6-acb6-39f60331c4ba)

CarinaWolli committed:
Merge branch 'main' into feat/schedule-timezone-change

Commit:
code clean up

Udit-takkar commented on packages/lib/date-ranges.ts:
```diff
@@ -10,25 +10,45 @@ export type DateRange = {
+function getAdjustedTimezone(date: Dayjs, timeZone: string, travelSchedules: TravelSchedule[]) {
+  let adjustedTimezone = timeZone;
+
+  travelSchedules.forEach((travelSchedule) => {
+    if (!date.isBefore(travelSchedule.startDate)) {
+      if (!travelSchedule.endDate || !date.isAfter(travelSchedule.endDate)) {
+        adjustedTimezone = travelSchedule.timeZone;
+      }
+    }
+  });
```
NIT

```suggestion
  travelSchedules.forEach((travelSchedule) => {
    if (!date.isBefore(travelSchedule.startDate) && (!travelSchedule.endDate || !date.isAfter(travelSchedule.endDate))  ) {
       
        adjustedTimezone = travelSchedule.timeZone;
      
    }
  });
```

Udit-takkar commented on apps/web/pages/settings/my-account/general.tsx:
```diff
@@ -181,6 +222,64 @@ const GeneralView = ({ localeProp, user }: GeneralViewProps) => {
+              {t("schedule_timezone_change")}
+            </Button>
+          ) : (
+            <div className="bg-muted border-subtle mt-2 rounded-md border p-4">
+              <Label>{t("travel_schedule")}</Label>
+              <div className="border-subtle mt-4 rounded-md border bg-white text-sm">
```
<img width="1512" alt="Screenshot 2024-03-06 at 6 23 07 PM" src="https://github.com/calcom/cal.com/assets/53316345/92185b94-e246-485a-8b41-dbfdca4752a8">

Dark mode

Udit-takkar commented on packages/features/schedules/components/DateOverrideList.tsx:
```diff
@@ -53,6 +58,7 @@ const DateOverrideList = ({
       new Date(end.toISOString().slice(0, -1))
     )}`;
   };
+  console.log(`items ${JSON.stringify(items)}`);
```
NIT: remove this

Udit-takkar commented on packages/lib/date-ranges.test.ts:
```diff
@@ -207,10 +207,85 @@ describe("processWorkingHours", () => {
+      })
+    ).toBeTruthy();
+
+    // 8AM in America/Kolkata is 2:30AM in UTC, 5PM in America/Kolkata is 11:30AM in UTC
```
```suggestion
    // 8AM in America/New_York is 2:30AM in UTC, 5PM in America/New_York is 11:30AM in UTC
```

Udit-takkar commented on packages/lib/date-ranges.test.ts:
```diff
@@ -207,10 +207,85 @@ describe("processWorkingHours", () => {
+      })
+    ).toBeTruthy();
+
+    // 8AM in America/Kolkata is 2:30AM in UTC, 5PM in America/Kolkata is 11:30AM in UTC
```
Asia/Kolkata or America/New_York?

CarinaWolli replied:
Oops, should be Asia/Kolkata. Fixed now

Udit-takkar commented on apps/web/components/settings/TravelScheduleModal.tsx:
```diff
@@ -0,0 +1,152 @@
+          <div className="mt-3">
+            <SettingsToggle
+              labelClassName="mt-1 font-normal"
+              title="Schedule timezone without end date"
```
NIT: use i18n

Commit:
code clean up from feedback

Commit:
fix asia/kalkota comment

Commit:
fix dark mode

Commit:
Merge branch 'main' into feat/schedule-timezone-change

Commit:
fix start and end date conversion to utc

Commit:
Merge branch 'main' into feat/schedule-timezone-change

emrysal committed:
Fix modal render issue

Commit:
Merge branch 'main' into feat/schedule-timezone-change

joeauyeung committed:
Merge branch 'main' into feat/schedule-timezone-change

joeauyeung commented on packages/features/schedules/components/DateOverrideList.tsx:
```diff
@@ -61,7 +66,17 @@ const DateOverrideList = ({
               item.ranges.map((range, i) => (
                 <p key={i} className="text-subtle text-xs">
-                  {timeSpan(range)}
+                  {`${timeSpan(range)} ${
+                    isDefaultSchedule
+                      ? travelSchedules.find(
+                          (travelSchedule) =>
+                            !dayjs(item.ranges[0].start).isBefore(travelSchedule.startDate) &&
+                            (!dayjs(item.ranges[0].end).isAfter(travelSchedule.endDate) ||
+                              !travelSchedule.endDate)
+                        )?.timeZone || ""
+                      : ""
```
Very nice considering this case.

joeauyeung commented on packages/lib/date-ranges.ts:
```diff
@@ -10,25 +10,46 @@ export type DateRange = {
+function getAdjustedTimezone(date: Dayjs, timeZone: string, travelSchedules: TravelSchedule[]) {
+  let adjustedTimezone = timeZone;
+
+  travelSchedules.forEach((travelSchedule) => {
```
Instead of looping through all the `travelSchedules` can we use a `.find()` or a `for ... of` loop and break once we find the schedule we're looking for

joeauyeung commented on packages/trpc/server/middlewares/sessionMiddleware.ts:
```diff
@@ -71,6 +71,7 @@ export async function getUserFromSession(ctx: TRPCContextInner, session: Maybe<S
       allowDynamicBooking: true,
       allowSEOIndexing: true,
       receiveMonthlyDigestEmail: true,
+      travelSchedules: true,
```
Why are we adding `travelSchedules` to the user session object? I would want to keep this as learn as possible.

CarinaWolli replied:
removed it and added a separate getTravelSchedules handler instead

joeauyeung commented on apps/web/pages/api/cron/changeTimeZone.ts:
```diff
@@ -0,0 +1,170 @@
+      });
+    }
+
+    await prisma.schedule.updateMany({
+      where: {
+        id: defaultScheduleId,
+      },
+      data: {
+        timeZone: timeZone,
+      },
+    });
```
I don't see any messaging about updating the default schedule timezone. I have two types of availabilities that I use for different event types. What if a user wants all their schedules updated? Could be a @ciaranha @itsberkaya question.

CarinaWolli replied:
@ciaranha and me talked about this a while ago. When changing the profile timezone only the default schedule should be updated. I think it actually would make sense to update the default schedule + every schedule that has the same timezone as the old profile timezone

joeauyeung commented on apps/web/pages/api/cron/changeTimeZone.ts:
Love seeing Dayjs being used like this here.

joeauyeung reviewed (changes requested):
Great work so far @CarinaWolli. I tried testing hitting the cron endpoint manually but I got the timezone to change but I think at a certain point I missed the time window. 

## Blocking
- On the profile it looks like the timezones aren't the translated strings (Hong_kong vs Hong Kong)
![image](https://github.com/calcom/cal.com/assets/65426560/8c004e7b-5765-459d-b87e-5b7e51e73190)

- ### Date Selector
	- When selecting a travel schedule with no end on dark mode the date selector is rendering weird
	![image](https://github.com/calcom/cal.com/assets/65426560/34f68991-6d51-4960-ada6-b110fae457f8)
	- Also with the date selector, the week starts on Monday but my settings are set to Sunday. This might cause some confusion for people.
	- I'm able to choose dates in the past
	- When adding multiple travel schedules, I think the dialog state should reset. (start and end dates, timezone)

- I set the new timezone to be today. When I created a booking it shows the booking time as the original timezone. I think if a user sets the current day for the new travel schedule it should update the bookings page. I'm wondering as well does the bookings page show the new updated times?
![image](https://github.com/calcom/cal.com/assets/65426560/fb7c7cb9-8dc8-4911-b106-49e8d12781a3)


## Non-blocking
- Would be cool to also schedule a timezone change here. Similar to how OOO is advertised
![image](https://github.com/calcom/cal.com/assets/65426560/29534614-3eef-429d-9f36-e5ac27e0b525)
- Should travel schedules be automatically added rather than hitting "Update" on the settings page? It just feels more intuitive to me. 
- It would be nice to edit individual travel schedules

Commit:
show timezone city wtihout _

Commit:
fix dark more for datepicker

Commit:
reset values after closing dialog

Commit:
remove session from middleware

Commit:
exit loop early once schedule is found

Commit:
fix type error

CarinaWolli committed:
Merge branch 'main' into feat/schedule-timezone-change

Commit:
add getTravelSchedules handler

Commit:
clean up DatePicker

CarinaWolli commented:
> * On the profile it looks like the timezones aren't the translated strings (Hong_kong vs Hong Kong)

 fixed, not translated but same as in the timezone select we just remove the _

> * ### Date Selector
>   
>   * When selecting a travel schedule with no end on dark mode the date selector is rendering weird

fixed
![Screenshot 2024-03-19 at 5 45 28 PM](https://github.com/calcom/cal.com/assets/30310907/1d2cde2d-a529-447d-9faf-7d309f0b4af8)

>   * Also with the date selector, the week starts on Monday but my settings are set to Sunday. This might cause some confusion for people.

this is an issue with DatePicker/DateRangePicker component let’s handle that in a separate PR

>   * I'm able to choose dates in the past

adding minDate which would solve that causes the weird dark mode design, that's also an issue with the DatePicker/DateRangePicker component, let’s handle that in a separate PR as well

>   * When adding multiple travel schedules, I think the dialog state should reset. (start and end dates, timezone)

fixed

> * I set the new timezone to be today. When I created a booking it shows the booking time as the original timezone. I think if a user sets the current day for the new travel schedule it should update the bookings page. I'm wondering as well does the bookings page show the new updated times?

When the cron job gets triggered your profile timezone would then get update and then everything including the bookings would show in the new timezone. You think we should check if the start date is today and if yes change the timezone right away?

CarinaWolli commented:
@joeauyeung I think I addressed all your blocking comments, let me know if I missed something. We can address the non-blocking comments in follow up

joeauyeung commented:
> > If I change the timezone today to a new one, I would want it to change now rather than wait until midnight. Could be another @ciaranha @itsberkaya question.
> 
> You wouldn't need to wait until midnight, we run the corn job on every full hour so it would trigger the next full hour. I just didn't see this as a use case of this feature to be honest, I would assume that the user would just change the timezone in the profile if they need it to be changed right away.

I was wondering if the person was already in the timezone and needed to change it on their profile immediately. It would be nice to give them this option to change it via the travel schedule since there's an end date. This is a non-blocking issue though.

Commit:
Merge branch 'main' into feat/schedule-timezone-change

Commit:
fix type error

CarinaWolli commented on apps/web/pages/availability/[schedule].tsx:
```diff
@@ -100,6 +105,7 @@ export default function Availability() {
             }
           : undefined
       }
+      travelSchedules={isDefaultSchedule ? travelSchedules || [] : []}
```
@calcom/platform I assume I also need to add the travelSchedules to PlatformAvailabilitySettingsWrapper now?

Commit:
Merge branch 'main' into feat/schedule-timezone-change

Commit:
code clean up

Commit:
code clean up

joeauyeung committed:
Merge branch 'main' into feat/schedule-timezone-change

ThyMinimalDev commented on apps/web/pages/api/cron/changeTimeZone.ts:
```diff
@@ -0,0 +1,170 @@
+    const endDateUTC = dayjs(travelSchedule.endDate).subtract(offset, "minute");
+    if (
+      !dayjs.utc().isBefore(startDateUTC) &&
+      dayjs.utc().isBefore(endDateUTC) &&
+      !travelSchedule.prevTimeZone
+    ) {
+      // if travel schedule has started and prevTimeZone is not yet set, we need to change time zone
+      await setNewTimeZone(travelSchedule.timeZone, travelSchedule.user);
+
+      if (!travelSchedule.endDate) {
+        await prisma.travelSchedule.delete({
+          where: {
+            id: travelSchedule.id,
+          },
+        });
+      } else {
+        await prisma.travelSchedule.update({
+          where: {
+            id: travelSchedule.id,
+          },
+          data: {
+            prevTimeZone: travelSchedule.user.timeZone,
+          },
+        });
+      }
+    }
+    if (!dayjs.utc().isBefore(endDateUTC)) {
+      if (travelSchedule.prevTimeZone) {
+        // travel schedule ended, change back to original timezone
+        await setNewTimeZone(travelSchedule.prevTimeZone, travelSchedule.user);
+      }
+      await prisma.travelSchedule.delete({
+        where: {
+          id: travelSchedule.id,
+        },
+      });
+    }
+  }
+  res.status(200).json({ timeZonesChanged });
+}
```
this cron is doing lots of individual calls to the database and might create heavy load on it, I suggest checking if we could use transactions and making sure the right indexes are defined in the schema

ThyMinimalDev replied:
now that I think about it, since it's a cron and it will most likely only execute once in a while, transaction is not important, only the indexes

CarinaWolli replied:
added indexes

ThyMinimalDev commented on apps/web/pages/api/cron/changeTimeZone.ts:
```diff
@@ -0,0 +1,170 @@
+    const startDateUTC = dayjs(travelSchedule.startDate).subtract(offset, "minute");
+    // 23:59 of user's time zone in utc time
+    const endDateUTC = dayjs(travelSchedule.endDate).subtract(offset, "minute");
+    if (
+      !dayjs.utc().isBefore(startDateUTC) &&
+      dayjs.utc().isBefore(endDateUTC) &&
+      !travelSchedule.prevTimeZone
+    ) {
+      // if travel schedule has started and prevTimeZone is not yet set, we need to change time zone
+      await setNewTimeZone(travelSchedule.timeZone, travelSchedule.user);
+
+      if (!travelSchedule.endDate) {
+        await prisma.travelSchedule.delete({
+          where: {
+            id: travelSchedule.id,
+          },
+        });
+      } else {
+        await prisma.travelSchedule.update({
+          where: {
+            id: travelSchedule.id,
+          },
+          data: {
+            prevTimeZone: travelSchedule.user.timeZone,
+          },
+        });
+      }
+    }
+    if (!dayjs.utc().isBefore(endDateUTC)) {
+      if (travelSchedule.prevTimeZone) {
+        // travel schedule ended, change back to original timezone
+        await setNewTimeZone(travelSchedule.prevTimeZone, travelSchedule.user);
+      }
+      await prisma.travelSchedule.delete({
+        where: {
+          id: travelSchedule.id,
+        },
+      });
+    }
+  }
```
we should avoid calling prisma inside of a for loop, can we here use the loop to aggregate the data to delete/update in vars and then do a deleteMany/updateMany after the for loop ?

CarinaWolli replied:
@ThyMinimalDev  added `deleteMany`, for the updates I think I would need to use transactions to optimize it. Should I do that or are we good like that?

joeauyeung commented:
@CarinaWolli looking through the commits since I last reviewed not much have changed so I'm ok to approve this PR once @ThyMinimalDev's comments have been addressed.

Commit:
add indexes

Commit:
use deleteMany

CarinaWolli commented on apps/web/pages/api/cron/changeTimeZone.ts:
```diff
@@ -0,0 +1,173 @@
+    }
+  }
+
+  await prisma.travelSchedule.deleteMany({
```
@ThyMinimalDev using delteMany now instead of deleting side for loop

Commit:
Merge branch 'main' into feat/schedule-timezone-change

CarinaWolli commented:
@ThyMinimalDev & @joeauyeung can you take a look again?

joeauyeung committed:
Merge branch 'main' into feat/schedule-timezone-change

joeauyeung reviewed (approved):
Looking at [`fd70b67` (#13951)](https://github.com/calcom/cal.com/pull/13951/commits/fd70b67e69d3d835ec2a5bf0b845596ea78aa118) and [`6f09c4c` (#13951)](https://github.com/calcom/cal.com/pull/13951/commits/6f09c4c1fefd1d42de162504bb92773321d397bf) seems like @ThyMinimalDev's comments have been addressed with no new changes to the functionality to be concerned about. Approving this PR.

Great job @CarinaWolli, I can't wait to use this feature myself!

joeauyeung reviewed (approved):
Looking at [`fd70b67` (#13951)](https://github.com/calcom/cal.com/pull/13951/commits/fd70b67e69d3d835ec2a5bf0b845596ea78aa118) and [`6f09c4c` (#13951)](https://github.com/calcom/cal.com/pull/13951/commits/6f09c4c1fefd1d42de162504bb92773321d397bf) seems like @ThyMinimalDev's comments have been addressed with no new changes to the functionality to be concerned about. Approving this PR.

Great job @CarinaWolli, I can't wait to use this feature myself!

CarinaWolli committed:
Merge branch 'main' into feat/schedule-timezone-change
